conn.go

```diff
@@ -510,10 +505,24 @@ var (
 func (c *Conn) exec(req frameWriter, tracer Tracer) (*framer, error) {
 	// TODO: move tracer onto conn
 	var stream int
-	select {
-	case stream = <-c.uniq:
-	case <-c.quit:
-		return nil, ErrConnectionClosed
+	maxStreams := 128 - 1
+	if c.cfg.ProtoVersion > protoVersion2 {
+		maxStreams = 32768 - 1
+	}
+	for stream == 0 {
+		if atomic.AddInt32(&c.availableStreams, -1) >= 0 {
+			stream = 1 + (((int(atomic.AddInt32(&c.streamCounter, 1)) % maxStreams) + maxStreams) % maxStreams)
```

Comment:
[Positive Modulus / Remainder function](http://stackoverflow.com/a/14997413/807701).

Since `streamCounter`/`stream` there is a possibility after a long enough time, `streamCounter` will be negative -> x & -n = -stream = panic index out of bounds, and we don't want that.